server/src/test/java/org/elasticsearch/index/query/TermsQueryBuilderTests.java

```diff
@@ -345,4 +327,155 @@ protected QueryBuilder parseQuery(XContentParser parser) throws IOException {
         return query;
     }
 
+    public void testSerializationFromLowerVersion() throws IOException {
+        {
+            TermsQueryBuilderBeforeV8 builder = new TermsQueryBuilderBeforeV8("foo", Arrays.asList(1, 3, 4), null);
+            TermsQueryBuilder copy = (TermsQueryBuilder) serializeAndDeserialize(Version.V_7_12_0, builder, Version.V_8_0_0, TermsQueryBuilder::new);
+            List<Object> values = copy.values();
+            assertEquals(Arrays.asList(1L, 3L, 4L), values);
+            assertTrue(copy.getValues() instanceof TermsQueryBuilder.ValuesBeforeV8);
+        }
+        {
+            TermsQueryBuilderBeforeV8 builder = new TermsQueryBuilderBeforeV8("foo", Arrays.asList(1L, 3L, 4L), null);
+            TermsQueryBuilder copy = (TermsQueryBuilder) serializeAndDeserialize(Version.V_7_12_0, builder, Version.V_8_0_0, TermsQueryBuilder::new);
+            List<Object> values = copy.values();
+            assertEquals(Arrays.asList(1L, 3L, 4L), values);
+            assertTrue(copy.getValues() instanceof TermsQueryBuilder.ValuesBeforeV8);
+        }
+        {
+            TermsQueryBuilderBeforeV8 builder = new TermsQueryBuilderBeforeV8("foo", Arrays.asList("a", "b", "c"), null);
+            TermsQueryBuilder copy = (TermsQueryBuilder) serializeAndDeserialize(Version.V_7_12_0, builder, Version.V_8_0_0, TermsQueryBuilder::new);
+            List<Object> values = copy.values();
+            assertEquals(Arrays.asList("a", "b", "c"), values);
+            assertTrue(copy.getValues() instanceof TermsQueryBuilder.ValuesBeforeV8);
+        }
+        {
+            TermsQueryBuilderBeforeV8 builder = new TermsQueryBuilderBeforeV8("foo", Arrays.asList(new BytesRef("a"), new BytesRef("b"), new BytesRef("c")), null);
+            TermsQueryBuilder copy = (TermsQueryBuilder) serializeAndDeserialize(Version.V_7_12_0, builder, Version.V_8_0_0, TermsQueryBuilder::new);
+            List<Object> values = copy.values();
+            assertEquals(Arrays.asList("a", "b", "c"), values);
+            assertTrue(copy.getValues() instanceof TermsQueryBuilder.ValuesBeforeV8);
+        }
+    }
+
+    public void testSerializationToLowerVersion() throws IOException {
```

Comment:
It would be easier to maintain if the version are randomized. You can use VersionUtils#randomCompatibleVersion(random, Version.CURRENT) to pick compatible versions randomly.
Otherwise we need to change the checks on every backport/new branch.

Reply:
I removed this method because this is also testing a serialization between different branches, as mentioned [here](https://github.com/elastic/elasticsearch/pull/67223#discussion_r557422972).